### PR TITLE
[#141778727]Framework agreement details expansion

### PR DESCRIPTION
**Pull request by risicle**
Story https://www.pivotaltracker.com/story/show/141778727

Ok, so, we're going to need to be able to actually *use* a whole load more keys in the `frameworkAgreementDetails` dictionary and pulling out specific ones and flattening them into the toplevel returned dictionary isn't going to cut it - at least having them returned under a specific `frameworkAgreementDetails` object gives us an amount of namespacing for these relatively obscure variables.

Another obstacle is that `frameworkAgreementDetails` itself has a json schema. Should we enforce the format of these new, obscure, variables or just open the schema to any `additionalProperties`? This strawman PR proposes the former.

Thoughts?

### CODE REVIEW

**Comment by risicle** [comment]
Further questions: in the schema I've just copied the formats of the date strings used in https://github.com/alphagov/digitalmarketplace-scripts/blob/01bf1272d0981d9725bbf951e82b46c25131518e/scripts/generate-framework-agreement-counterpart-signature-pages.py#L14 though I'm really thinking if we're going to be specifying these we should probably be doing so in a standardized format and not just the string-format that happens to get used in the place ...where it's used.

File: app/models.py [hunk 1]
```diff
@@ -124,16 +124,19 @@ def serialize(self):
...
-            'frameworkAgreementVersion': (self.framework_agreement_details or {}).get("frameworkAgreementVersion"),
-            'variations': (self.framework_agreement_details or {}).get("variations", {}),
             'status': self.status,
             'clarificationQuestionsOpen': self.clarification_questions_open,
             'lots': [lot.serialize() for lot in self.lots],
             'application_close_date': (
                 self.application_close_date and self.application_close_date.strftime(DATETIME_FORMAT)
             ),
-            'allow_declaration_reuse': self.allow_declaration_reuse
+            'allow_declaration_reuse': self.allow_declaration_reuse,
```
Comment (allait) [inline app/models.py]: I don't know why the field above is breaking the convention, but JSON response keys should be in camel-case.

File: json_schemas/framework-agreement-details.json [hunk 2]
```diff
@@ -15,6 +15,57 @@
       "minLength": 1,
       "type": "string"
     },
+    "frameworkRefDate": {
```
Comment (allait) [inline json_schemas/framework-agreement-details.json]: What does 'Ref' stand for? Can we expand this key name or is it already used somewhere?

File: json_schemas/framework-agreement-details.json [hunk 3]
```diff
@@ -15,6 +15,57 @@
...
+      "minLength": 1,
+      "type": "string",
+      "patterm": "^[0-9-]+$"
+    },
+    "frameworkStartDate": {
```
Comment (allait) [inline json_schemas/framework-agreement-details.json]: I feel like these things should be part of the Framework itself. I think @benvand was looking into this at some point.

File: json_schemas/framework-agreement-details.json [hunk 4]
```diff
@@ -15,6 +15,57 @@
...
+      "minLength": 1,
+      "type": "string",
+      "patterm": "^[0-9/]+$"
+    },
+    "frameworkEndDate": {
+      "minLength": 1,
+      "type": "string",
+      "patterm": "^[0-9/]+$"
```
Comment (allait) [inline json_schemas/framework-agreement-details.json]: patter**n**

Comment (allait) [inline json_schemas/framework-agreement-details.json]: Also, maybe this should be using the built-in `date-time` format (https://spacetelescope.github.io/understanding-json-schema/reference/string.html)

File: app/models.py (hunk 1 above)
Comment (risicle) [inline app/models.py]: I noticed this too but that's not part of this PR - think this is just a trailing comma that got added.

File: json_schemas/framework-agreement-details.json (hunk 2 above)
Comment (risicle) [inline json_schemas/framework-agreement-details.json]: good point - @TheDoubleK - this was just transposed from the script - thoughts on better naming?

File: json_schemas/framework-agreement-details.json (hunk 3 above)
Comment (risicle) [inline json_schemas/framework-agreement-details.json]: If you're reasonably sure we won't want to be able to override these dates just for the purposes of output on the signature page (@TheDoubleK wasn't so sure) I'd agree.

File: json_schemas/framework-agreement-details.json (hunk 4 above)
Comment (risicle) [inline json_schemas/framework-agreement-details.json]: goddamn, and I probably agree.

File: app/models.py (hunk 1 above)
Comment (allait) [inline app/models.py]: Ah, of course, my bad. Never mind then.

File: json_schemas/framework-agreement-details.json [hunk 5]
```diff
@@ -15,6 +15,57 @@
       "minLength": 1,
       "type": "string"
     },
+    "frameworkRefDate": {
+      "minLength": 1,
+      "type": "string",
+      "patterm": "^[0-9-]+$"
```
Comment (TheDoubleK) [inline json_schemas/framework-agreement-details.json]: Typo: `patterm`

File: json_schemas/framework-agreement-details.json (hunk 2 above)
Comment (TheDoubleK) [inline json_schemas/framework-agreement-details.json]: It's a date that becomes part of the document reference on the framework agreement. It doesn't really make any sense outside of the context of being a thing that goes into the placeholder.  It's formatted differently from dates that go in to the text as dates rather than as part of the document reference number.

Comment (TheDoubleK) [inline json_schemas/framework-agreement-details.json]: So to answer your question `dateForframeworkAgreementReferenceNumber` would be more descriptive, but I'm not sure it really helps :/

File: json_schemas/framework-agreement-details.json (hunk 3 above)
Comment (TheDoubleK) [inline json_schemas/framework-agreement-details.json]: I'd prefer to keep this framework agreement data dumb - i.e. it's just a collection of the exact strings we want to patch into the agreement document template.  

It feels like overkill to start saving actual timestamps and having the script to generate agreements know where to find the various timestamps and how they should be formatted for different placeholders in the template.  And what if the format changes from one framework agreement to the next?  We still need to be able to generate the right thing for all frameworks.  Just storing the strings in a convenient blob feels like the best thing.

File: json_schemas/framework-agreement-details.json (hunk 4 above)
Comment (TheDoubleK) [inline json_schemas/framework-agreement-details.json]: I don't think it should - we don't want an RFC compliant date.  We want the exact string as it should appear in the framework agreement document.  Arguably we shouldn't enforce any particular formatting here - just a non-empty string.

Comment (TheDoubleK) [inline json_schemas/framework-agreement-details.json]: Actually yes, I'd argue that we don't want to specify any regex or format for any of these strings.  These are just strings to go into placeholders, we don't care in the DB what format they are.

Comment (risicle) [inline json_schemas/framework-agreement-details.json]: Valid opinion. Thoughts @allait ?

Comment (galund) [inline json_schemas/framework-agreement-details.json]: FWIW, I think the highest priority would be to avoid duplication (opportunity for inconsistency) of data. Ideally, all these dates would be in some non-ambiguous machine-readable format, formatted for display on the front end, and 'human readable date strings' in the framework data would _not_ be a thing.

Comment (TheDoubleK) [inline json_schemas/framework-agreement-details.json]: * There is no duplication here at all - we don't store these dates anywhere else in the database. (We do store the "applications closed" date, but that is none of the dates here.)

* The data here (i.e. in the frameworkAgreementDetails block) is *not* intended to be used anywhere at all on the frontend.  It is for use by the scripts that generate the signature page PDFs.

* The main requirement of those PDFs is that they *exactly* match the page in the full framework agreement document, so storing the *exact* strings that need to go in there makes sense.

* I'm not sure in what sense these are ambiguous at all?

**Comment by TheDoubleK** [comment]
I think it's worth explaining the exact intent of these changes.

We have two scripts that we use to generate framework agreement pages (unsigned and countersigned versions).

Currently these scripts take a bunch of data at the command line from a local JSON file - data that is needed to pass to the template for the agreement signature page with the right framework name, document reference number, etc. - e.g. see the docstring here for details:
See here: https://github.com/alphagov/digitalmarketplace-scripts/blob/master/scripts/generate-framework-agreement-signature-pages.py

That was cobbled together so we could quickly make the DOS2 agreement pages, but there is no "source of truth" about what those values need to be for making the documents.  So this is intended to have the required details for the agreement in the database rather than in a file on my machine that I put together with Tobi at the last minute.

I strongly feel that these "generate agreements" scrips should not have to know about how to format dates for different fields.  They should just get the exact strings they need to patch into the agreement page template.  Doing things this way also means we can concurrently support different formats for reference numbers, dates, etc. on the page: we just need to make sure the per-framework details are right and then we can mix it up between frameworks as much as we like.  (I'm not saying we should be mixing it up, but you know the whims of lawyers, designers, etc...)

**Comment by allait** [comment]
@TheDoubleK yeah, I think that makes sense. I misread the PR title and was under the impression that we're storing a copy of this on each `frameworkAgrement` before I realised this is a field on the Framework model.

I agree with @galund that consistency is a concern. We've seen the unpleasant effects of small variations in framework data when it's not coming from a single source (like the lot numbering in the agreement letters for example) and I think that's something we need to keep in mind.

Ideally, we'd have these dates stored on a framework and then the API could add the keys to the agreement details to present the same data in a different format. But that's something we can do in the future once we have an actual use for these dates somewhere else and understand a bit more how these formats could change in the agreements.

So if @TheDoubleK is happy with this it LGTM.

**Comment by TheDoubleK** [comment]
Yep works for me :+1:

**Comment by TheDoubleK** [comment]
I've put a bug in our Icebox about the camelCase keys: https://www.pivotaltracker.com/story/show/141950783

**Comment by risicle** [comment]
I'll remove the `parrern` restrictions on the date fields before merging.
